### PR TITLE
is_const_array and Distinct with &[impl Borrow<Self>]

### PR DESCRIPTION
**Pull request by Pat-Lafon**
- Distinct currently takes `&[&Self]` for which I would like to do a similar api relaxing like in #246.
- With the new Rust version, there is a minor clippy change.
- Python has this nice `is_const_array(a)` function which I would like to propose here.

### CODE REVIEW

File: z3/src/ast.rs [hunk 1]
```diff
@@ -1525,6 +1528,26 @@ impl<'ctx> Array<'ctx> {
             })
         }
     }
+
+    /// Returns true if the array is a const array(a.is_const_array() <=> exists v, forall i. select(a, i) == v)
```
Comment (waywardmonkeys) [inline z3/src/ast.rs:1532]: Is this correct?

**Review by waywardmonkeys — CHANGES_REQUESTED** [review]

File: z3/src/ast.rs (hunk 1 above)
Comment (Pat-Lafon) [reply z3/src/ast.rs:1532]: That is coming from https://github.com/prove-rs/z3.rs/blob/3c9766336297769467be0ddcbce0376b58a344f3/z3-sys/src/lib.rs#L520-L522. So I believe so? Maybe only a single implication now that I think of it because I'm not sure if a nested store gets collapsed. (store (store ((as const (Array Int Int)) 9) 0 1281) 0 9)

Comment (waywardmonkeys) [reply z3/src/ast.rs]: Sorry, I had commented while doing something else and from my phone away from home.

My concern was actually about the formatting and if that should be in the summary line at the top of the comment.

Should also have a blank line before the example header.

File: z3/tests/lib.rs [hunk 2]
```diff
@@ -1221,7 +1221,7 @@ fn test_tactic_fail() {
 
     let tactic = Tactic::new(&ctx, "fail");
     let apply_results = tactic.apply(&goal, Some(&params));
-    assert!(matches!(apply_results, Err(_)));
+    assert!(apply_results.is_err());
```
Comment (waywardmonkeys) [inline z3/tests/lib.rs:1224]: I've landed this separately (also done by another PR). This commit can be removed from this PR.

**Comment by waywardmonkeys** [comment]
Could you squash the various commits for `is_const_array` into a single commit? After removing the clippy fix (already on `master`), that would leave one commit for the `Distinct` change and one for `is_const_array`.
